RecoMTD/TrackExtender/plugins/TrackExtenderWithMTD.cc

```diff
@@ -207,7 +207,87 @@ namespace {
     }
     return tofpid;
   }
+
+  bool getTrajectoryStateClosestToBeamLine(const Trajectory& traj,
+                                           const reco::BeamSpot& bs,
+                                           const Propagator* thePropagator,
+                                           TrajectoryStateClosestToBeamLine& tscbl) {
+    // get the state closest to the beamline
+    TrajectoryStateOnSurface stateForProjectionToBeamLineOnSurface =
+        traj.closestMeasurement(GlobalPoint(bs.x0(), bs.y0(), bs.z0())).updatedState();
+
+    if (!stateForProjectionToBeamLineOnSurface.isValid()) {
+      edm::LogError("CannotPropagateToBeamLine") << "the state on the closest measurement isnot valid. skipping track.";
+      return false;
+    }
+
+    const FreeTrajectoryState& stateForProjectionToBeamLine = *stateForProjectionToBeamLineOnSurface.freeState();
+
+    TSCBLBuilderWithPropagator tscblBuilder(*thePropagator);
+    tscbl = tscblBuilder(stateForProjectionToBeamLine, bs);
+
+    return tscbl.isValid();
+  }
+
+  bool trackPathLength(const Trajectory& traj,
+                       const TrajectoryStateClosestToBeamLine& tscbl,
+                       const Propagator* thePropagator,
+                       double& pathlength) {
```

Comment:
sorry for making this question, but I have a doubt :(
do we really need `double` precision here ?

Reply:
I suspect no, but it will change a numerical result for the PR which I wanted to stay technical.